src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+        const section = parametersWithVerb.values.find((k: { name: string; }) => {
+          return k.name === `$${param}`;
+        });
+
+        let list: string[] = [];
+        if (section && section.items && section.items.length > 0) {
+          list = section.items;
+          this.setState({
+            filteredSuggestions: list,
+            suggestions: list,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  private filterSuggestions(userInput: string, previousUserInput: string, compare: string, suggestions: string[]) {
+    let compareString = userInput.replace(previousUserInput, '');
+    if (compare) {
+      compareString = compare + compareString;
+    }
+    // Filter our suggestions that don't contain the user's input
+    const filteredSuggestions = suggestions.filter((suggestion: string) => {
+      return suggestion.toLowerCase().indexOf(compareString.toLowerCase()) > -1;
+    });
+    this.setState({
+      filteredSuggestions,
+      compare: compareString
+    });
+  }
+
+  private requestForAutocompleteOptions(url: string) {
+    const { requestUrl, queryVersion } = parseSampleUrl(url);
+    if (requestUrl || queryVersion) {
+      if (!this.props.autoCompleteOptions || `${requestUrl}` !== this.props.autoCompleteOptions.url) {
+        this.props.actions!.fetchAutoCompleteOptions(requestUrl);
+      }
+      else {
+        this.performLocalSearch(url);
+      }
+    }
+  }
+
+  private performLocalSearch(url: string) {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '/':
+        this.displayLinkOptions();
+        break;
+
+      case '?':
+        this.getQueryParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  private appendSuggestionToUrl(selected: string) {
+    const { userInput, compare } = this.state;
+    const selectedSuggestion = cleanUpSelectedSuggestion(compare, userInput, selected);
+    this.setState({
+      activeSuggestion: 0,
+      filteredSuggestions: [],
+      showSuggestions: false,
+      userInput: selectedSuggestion,
+      compare: ''
+    });
+    this.props.contentChanged(selectedSuggestion);
+    this.setFocus();
+  }
+
+  public render() {
+    const {
+      activeSuggestion,
+      filteredSuggestions,
+      showSuggestions,
+      userInput
+    } = this.state;
+
+    const { fetchingSuggestions, sampleQuery } = this.props;
+
+    const currentTheme = getTheme();
+    const suggestionClass: any = queryInputStyles(currentTheme).autoComplete.suggestions;
+    const suggestionOption: any = queryInputStyles(currentTheme).autoComplete.suggestionOption;
+    const activeSuggestionClass: any = queryInputStyles(currentTheme).autoComplete.suggestionActive;
+    const autoInput: any = queryInputStyles(currentTheme).autoComplete.input;
+    const suggestionTitle: any = queryInputStyles(currentTheme).autoComplete.suggestionTitle;
```

Comment:
To avoid calling queryInputStyles 5 times in a row:
```suggestion
  const { suggestions : suggestionClass,
        suggestionOption,
        suggestionActive : activeSuggestionClass,
        input : autoInput,
        suggestionTitle } = queryInputStyles(currentTheme).autoComplete;
```